torch/optim/adagrad.py

```diff
@@ -55,9 +57,30 @@ def __init__(
             foreach=foreach,
             maximize=maximize,
             differentiable=differentiable,
+            fused=fused,
         )
         super().__init__(params, defaults)
 
+        if fused:
+            if differentiable:
+                raise RuntimeError("`fused` does not support `differentiable`")
+            self._step_supports_amp_scaling = True
+            fused_supported_devices = _get_fused_kernels_supported_devices()
+            # Not support CUDA yet
+            cuda = fused_supported_devices.pop(0)
+            assert cuda == "cuda"
+            if not all(
+                p.device.type in fused_supported_devices and torch.is_floating_point(p)
+                for pg in self.param_groups
+                for p in pg["params"]
+            ):
+                raise RuntimeError(
+                    "`fused=True` requires all the params to be floating point Tensors of "
+                    f"supported devices: {fused_supported_devices}."
+                )
+            if foreach:
+                raise RuntimeError("`fused` and `foreach` cannot be `True` together.")
+
```

Comment:
in line 87 below, does step need to be init'd different for fused adagrad?

Reply:
Thanks for advice, updated.